### PR TITLE
ddl: fix create and alter check constraints problems 

### LINKED ISSUE #44737: Whether the intermediate state setting of `alter constraint` is appropriate
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
When we using `alter constraint` to switch from `not enforced` to `enforced`, we may not see the `constraint` information when executing `show create table`.
https://github.com/pingcap/tidb/blob/052c17f4f69bdedd8fae91c75e95cb70a5c6385f/ddl/constraint.go#L216-L242

### 2. What did you expect to see? (Required)
see the `constraint` information

### 3. What did you see instead (Required)
 may not see the `constraint` information

### 4. What is your TiDB version? (Required)

master
After https://github.com/pingcap/tidb/pull/44455

### LINKED ISSUE #47567: CHECK CONSTRAINT loss when ENFORCED again
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
mysql> show create table t\G
*************************** 1. row ***************************
       Table: t
Create Table: CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL,
CONSTRAINT `t_chk_1` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */,
CONSTRAINT `t_chk_2` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */,
CONSTRAINT `t_chk_3` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */,
CONSTRAINT `check_1` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */,
CONSTRAINT `t_chk_4` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin
1 row in set (0.00 sec)

mysql> insert t select 1;
Query OK, 1 row affected (0.00 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> select * from t;
+------+
| a    |
+------+
|    2 |
|    2 |
|    1 |
+------+
3 rows in set (0.01 sec)

mysql> ALTER TABLE t ALTER CONSTRAINT t_chk_1 ENFORCED;
ERROR 3819 (HY000): Check constraint 't_chk_1' is violated.
mysql> select * from t;
+------+
| a    |
+------+
|    2 |
|    2 |
|    1 |
+------+
3 rows in set (0.01 sec)

mysql> show create table t\G
*************************** 1. row ***************************
       Table: t
Create Table: CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL,
CONSTRAINT `t_chk_2` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */,
CONSTRAINT `t_chk_3` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */,
CONSTRAINT `check_1` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */,
CONSTRAINT `t_chk_4` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin
1 row in set (0.00 sec)

mysql> select * from information_schema.CHECK_CONSTRAINTS;
+--------------------+-------------------+-----------------+--------------+
| CONSTRAINT_CATALOG | CONSTRAINT_SCHEMA | CONSTRAINT_NAME | CHECK_CLAUSE |
+--------------------+-------------------+-----------------+--------------+
| def                | sbtest            | check_1         | (1 < `a`)    |
| def                | sbtest            | t_chk_2         | (1 < `a`)    |
| def                | sbtest            | t_chk_3         | (1 < `a`)    |
| def                | sbtest            | check_1         | (1 < `a`)    |
| def                | sbtest            | t_chk_4         | (1 < `a`)    |
+--------------------+-------------------+-----------------+--------------+
5 rows in set (0.00 sec)

mysql>
```

### 2. What did you expect to see? (Required)

`t_chk_1` should be keep NOT ENFORCED, but not loss.

and, in mysql, the logic is right.
```
mysql> select version();
+-----------+
| version() |
+-----------+
| 8.1.0     |
+-----------+
1 row in set (0.00 sec)

mysql> ALTER TABLE t ALTER CONSTRAINT t_chk_1 ENFORCED;
ERROR 3819 (HY000): Check constraint 't_chk_1' is violated.
mysql> select * from t;
+------+
| a    |
+------+
|    1 |
|    2 |
|    2 |
+------+
3 rows in set (0.00 sec)

mysql> show create table t\G
*************************** 1. row ***************************
       Table: t
Create Table: CREATE TABLE `t` (
  `a` int DEFAULT NULL,
  CONSTRAINT `check_1` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */,
  CONSTRAINT `t_chk_1` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */,
  CONSTRAINT `t_chk_2` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */,
  CONSTRAINT `t_chk_3` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */,
  CONSTRAINT `t_chk_4` CHECK ((1 < `a`)) /*!80016 NOT ENFORCED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb3 COLLATE=utf8mb3_bin
1 row in set (0.00 sec)

mysql> select * from information_schema.CHECK_CONSTRAINTS;
+--------------------+-------------------+-----------------+--------------+
| CONSTRAINT_CATALOG | CONSTRAINT_SCHEMA | CONSTRAINT_NAME | CHECK_CLAUSE |
+--------------------+-------------------+-----------------+--------------+
| def                | sbtest            | t_chk_1         | (1 < `a`)    |
| def                | sbtest            | t_chk_2         | (1 < `a`)    |
| def                | sbtest            | t_chk_3         | (1 < `a`)    |
| def                | sbtest            | check_1         | (1 < `a`)    |
| def                | sbtest            | t_chk_4         | (1 < `a`)    |
+--------------------+-------------------+-----------------+--------------+
5 rows in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

Release Version: v7.5.0-alpha

### LINKED ISSUE #47631: Add not enforced check not compatible with mysql 8.0
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
CREATE TABLE `t` (`a` int(11) DEFAULT NULL);
insert t values(1);
alter table t ADD CONSTRAINT chk CHECK (a > 1) NOT ENFORCED;
```

### 2. What did you expect to see? (Required)

```
MySQL 8.0:

mysql> alter table t ADD CONSTRAINT chk CHECK (a > 1) NOT ENFORCED;
Query OK, 0 rows affected (0.02 sec)
Records: 0  Duplicates: 0  Warnings: 0
```

### 3. What did you see instead (Required)

```

mysql> alter table t ADD CONSTRAINT chk CHECK (a > 1) NOT ENFORCED;
ERROR 3819 (HY000): Check constraint 'chk' is violated.

```

### 4. What is your TiDB version? (Required)

master

### LINKED ISSUE #47632: check constraints ddl can't return forever
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
CREATE TABLE `t` (`a` int(11) DEFAULT NULL);
insert t values(1);
alter table t ADD CONSTRAINT chk CHECK (a > 1) ENFORCED;
alter table t ADD CONSTRAINT chk CHECK (a > 1) ENFORCED;
alter table t drop CONSTRAINT chk;
```

### 2. What did you expect to see? (Required)

```
MySQL 8.0:

mysql> CREATE TABLE `t` (`a` int(11) DEFAULT NULL);
Query OK, 0 rows affected, 1 warning (0.04 sec)

mysql> insert t values(1);
Query OK, 1 row affected (0.01 sec)

mysql> alter table t ADD CONSTRAINT chk CHECK (a > 1) ENFORCED;
ERROR 3819 (HY000): Check constraint 'chk' is violated.
mysql> alter table t ADD CONSTRAINT chk CHECK (a > 1) ENFORCED;
ERROR 3819 (HY000): Check constraint 'chk' is violated.
mysql> alter table t drop CONSTRAINT chk;
ERROR 3940 (HY000): Constraint 'chk' does not exist.

```

### 3. What did you see instead (Required)

```

mysql> CREATE TABLE `t` (`a` int(11) DEFAULT NULL);
Query OK, 0 rows affected (0.12 sec)

mysql> insert t values(1);
Query OK, 1 row affected (0.00 sec)

mysql> alter table t ADD CONSTRAINT chk CHECK (a > 1) ENFORCED;
ERROR 3819 (HY000): Check constraint 'chk' is violated.
mysql> alter table t ADD CONSTRAINT chk CHECK (a > 1) ENFORCED;
ERROR 3822 (HY000): Duplicate check constraint name 'chk'.
mysql> alter table t drop CONSTRAINT chk;
----------can't return forever------------

TiDB errlog:

[2023/10/14 10:16:29.470 +08:00] [INFO] [ddl_worker.go:996] ["run DDL job"] [worker="worker 1, tp general"] [category=ddl] [jobID=203] [conn=2048917508] [category=ddl] [job="ID:203, Type:drop check constraint, State:cancelling, SchemaState:none, SchemaID:2, TableID:200, RowCount:0, ArgLen:0, start time: 2023-10-14 10:12:26.329 +0800 CST, Err:[ddl:8204]Invalid DDL job%!(EXTRA string=constraint, model.SchemaState=public), ErrCount:237, SnapshotVersion:0"]
[2023/10/14 10:16:29.474 +08:00] [INFO] [ddl_worker.go:1207] ["schema version doesn't change"] [category=ddl]
[2023/10/14 10:16:29.481 +08:00] [INFO] [ddl_worker.go:996] ["run DDL job"] [worker="worker 1, tp general"] [category=ddl] [jobID=203] [conn=2048917508] [category=ddl] [job="ID:203, Type:drop check constraint, State:running, SchemaState:none, SchemaID:2, TableID:200, RowCount:0, ArgLen:0, start time: 2023-10-14 10:12:26.329 +0800 CST, Err:[ddl:8204]Invalid DDL job%!(EXTRA string=constraint, model.SchemaState=public), ErrCount:237, SnapshotVersion:0"]
[2023/10/14 10:16:29.483 +08:00] [WARN] [ddl_worker.go:958] ["run DDL job error"] [worker="worker 1, tp general"] [category=ddl] [jobID=203] [conn=2048917508] [error="[ddl:8204]Invalid DDL job%!(EXTRA string=constraint, model.SchemaState=public)"]
[2023/10/14 10:16:29.485 +08:00] [WARN] [ddl_worker.go:966] ["DDL job error count exceed the limit, cancelling it now"] [worker="worker 1, tp general"] [category=ddl] [jobID=203] [conn=2048917508] [errorCountLimit=3]
[2023/10/14 10:16:29.491 +08:00] [INFO] [ddl_worker.go:840] ["run DDL job failed, sleeps a while then retries it."] [worker="worker 1, tp general"] [category=ddl] [jobID=203] [conn=2048917508] [waitTime=1s] [error="[ddl:8204]Invalid DDL job%!(EXTRA string=constraint, model.SchemaState=public)"]

```

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by jiyfhust**
### What problem does this PR solve?

Issue Number: close #47567 close #47631 close #47632

It seems also close #44737

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/constraint_test.go [hunk 1]
```diff
@@ -287,3 +287,21 @@ func TestAlterEnforcedConstraintStateChange(t *testing.T) {
...
+// Related issue TiDB#47567, #47631 and #47632.
+func TestCheckConstraintIssue47567(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+
+	tk.MustExec("set @@global.tidb_enable_check_constraint = 1")
+	tk.MustExec("use test")
+	tk.MustExec("CREATE TABLE `t` (`a` int(11) DEFAULT NULL)")
+	tk.MustExec("insert t values(1)")
+	tk.MustGetErrMsg("alter table t ADD CONSTRAINT chk CHECK (a > 1) ENFORCED", "[ddl:3819]Check constraint 'chk' is violated.")
```
Comment (CbcWestwolf) [inline pkg/ddl/constraint_test.go:300]: Why is there two same SQLs here?

Comment (jiyfhust) [reply pkg/ddl/constraint_test.go:300]: As issue #47632 reproduce steps，the first sql will rollback failed, and the second will report "Duplicate check constraint name" error.

File: pkg/ddl/constraint_test.go [hunk 2]
```diff
@@ -287,3 +287,21 @@ func TestAlterEnforcedConstraintStateChange(t *testing.T) {
 	tk.MustExec("alter table t alter constraint c1 enforced")
 	tk.MustQuery("select * from t").Check(testkit.Rows("12"))
 }
+
+// Related issue TiDB#47567, #47631 and #47632.
+func TestCheckConstraintIssue47567(t *testing.T) {
```
Comment (CbcWestwolf) [inline pkg/ddl/constraint_test.go:292]: We can move this test case to https://github.com/pingcap/tidb/blob/master/tests/integrationtest/t/ddl/constraint.test#L660, and run it by 
```bash
cd tests/integrationtest
bash ./run-tests.sh -s ../../bin/tidb-server -r ddl/constraint
```

File: pkg/ddl/constraint_test.go (hunk 1 above)
Comment (CbcWestwolf) [reply pkg/ddl/constraint_test.go:300]: Got it!

File: pkg/ddl/constraint_test.go (hunk 2 above)
Comment (jiyfhust) [reply pkg/ddl/constraint_test.go]: Tests have been moved.

File: tests/integrationtest/t/ddl/constraint.test [hunk 3]
```diff
@@ -658,4 +658,23 @@ show create table t;
...
+drop table if exists t;
+set @@global.tidb_enable_check_constraint = 1;
+CREATE TABLE `t` (`a` int(11) DEFAULT NULL);
+show create table t;
+insert t values(1);
+select * from t;
+-- error 3940
+alter table t ADD CONSTRAINT chk CHECK (a > 1) ENFORCED;
+-- error 3940
+alter table t ADD CONSTRAINT chk CHECK (a > 1) ENFORCED;
```
Comment (tiancaiamao) [inline tests/integrationtest/t/ddl/constraint.test:672]: duplicated with line 669-670?

Comment (jiyfhust) [reply tests/integrationtest/t/ddl/constraint.test:672]: No, it is desired. https://github.com/pingcap/tidb/pull/47633#discussion_r1375588267 
Same as it.

Comment (tiancaiamao) [reply tests/integrationtest/t/ddl/constraint.test:672]: > issue #47632 reproduce steps，the first sql will rollback failed, and the second will report "Duplicate check constraint name" error.

Get it

**Review by tiancaiamao — APPROVED** [review]

**Comment by tiancaiamao** [comment]
@CbcWestwolf PTAL

**Review by CbcWestwolf — APPROVED** [review]
LGTM

**Comment by jiyfhust** [comment]
Hi, @hawkingrei How can we finish the check-issue-triage-complete?

**Comment by jiyfhust** [comment]
PTAL @tiancaiamao

**Review by tangenta — APPROVED** [review]

**Pull request merged**